ledger/src/shred/merkle.rs

```diff
@@ -0,0 +1,522 @@
+        }
+        let proof_size = self.proof_size()?;
+        let shard_size = Self::size_of_erasure_encoded_slice(proof_size)?;
+        self.payload
+            .get(SIZE_OF_CODING_SHRED_HEADERS..SIZE_OF_CODING_SHRED_HEADERS + shard_size)
+            .ok_or(Error::InvalidPayloadSize(self.payload.len()))
+    }
+
+    fn sanitize(&self) -> Result<(), Error> {
+        match self.common_header.shred_variant {
+            ShredVariant::MerkleCode(proof_size) => {
+                if self.merkle_branch.proof.len() != usize::from(proof_size) {
+                    return Err(Error::InvalidProofSize(proof_size));
+                }
+            }
+            _ => return Err(Error::InvalidShredVariant),
+        }
+        if !self.verify_merkle_proof()? {
+            return Err(Error::InvalidMerkleProof);
+        }
+        shred_code::sanitize(self)
+    }
+
+    fn signed_message(&self) -> &[u8] {
+        self.merkle_branch.root.as_ref()
+    }
+}
+
+impl ShredDataTrait for ShredData {
+    #[inline]
+    fn data_header(&self) -> &DataShredHeader {
+        &self.data_header
+    }
+
+    fn data(&self) -> Result<&[u8], Error> {
+        let proof_size = self.proof_size()?;
+        let data_buffer_size = Self::capacity(proof_size)?;
+        let size = usize::from(self.data_header.size);
+        if size > self.payload.len()
+            || size < SIZE_OF_DATA_SHRED_HEADERS
+            || size > SIZE_OF_DATA_SHRED_HEADERS + data_buffer_size
+        {
+            return Err(Error::InvalidDataSize {
+                size: self.data_header.size,
+                payload: self.payload.len(),
+            });
+        }
+        Ok(self.payload[SIZE_OF_DATA_SHRED_HEADERS..size])
+    }
+
+    // Only for tests.
+    fn set_last_in_slot(&mut self) {
+        self.data_header.flags |= ShredFlags::LAST_SHRED_IN_SLOT;
+        let buffer = &mut self.payload[SIZE_OF_COMMON_SHRED_HEADER..];
+        bincode::serialize_into(buffer, &self.data_header).unwrap();
+    }
+}
+
+impl ShredCodeTrait for ShredCode {
+    #[inline]
+    fn coding_header(&self) -> &CodingShredHeader {
+        &self.coding_header
+    }
+}
+
+// Obtains parent's hash by joining two sibiling nodes in merkle tree.
+fn join_nodes<S: AsRef<[u8]>, T: AsRef<[u8]>>(node: S, other: T) -> Hash {
+    let node = &node.as_ref()[..SIZE_OF_MERKLE_PROOF_ENTRY];
+    let other = &other.as_ref()[..SIZE_OF_MERKLE_PROOF_ENTRY];
+    hashv(&[MERKLE_HASH_PREFIX_NODE, node, other])
+}
+
+fn verify_merkle_proof(index: usize, node: Hash, merkle_branch: &MerkleBranch) -> bool {
+    let proof = merkle_branch.proof.iter();
+    proof.fold((index, node), |(index, node), other| {
+        let parent = if index % 2 == 0 {
+            join_nodes(node, other)
+        } else {
+            join_nodes(other, node)
+        };
+        (index >> 1, parent)
+    }) == (0usize, merkle_branch.root)
+}
+
+#[cfg(test)]
+fn make_merkle_tree(mut nodes: Vec<Hash>) -> Vec<Hash> {
```

Comment:
Temporarily marked as `#[cfg(test)]` to avoid clippy dead-code errors;
but this code will be used in shreds construction.